packages/myst-toc/src/toc.ts

```diff
@@ -1,21 +1,193 @@
-import type { TOC } from './types.js';
-import schema from './schema.json';
-import _Ajv from 'ajv';
+import type { TOC, Entry, FileEntry, URLEntry, PatternEntry, ParentEntry } from './types.js';
+
+import type { ValidationOptions } from 'simple-validators';
+import {
+  defined,
+  incrementOptions,
+  validateBoolean,
+  validateEnum,
+  validateList,
+  validateNumber,
+  validateObjectKeys,
+  validateString,
+  validationError,
+} from 'simple-validators';
+
+const COMMON_ENTRY_KEYS = ['title', 'hidden', 'numbering', 'id', 'part', 'class'];
+
+function validateCommonEntry(entry: any, opts: ValidationOptions): Entry | undefined {
+  if (defined(entry.title)) {
+    entry.title = validateString(entry.title, incrementOptions('title', opts));
+  }
+
+  if (defined(entry.hidden)) {
+    entry.hidden = validateBoolean(entry.hidden, incrementOptions('hidden', opts));
+  }
+
+  if (defined(entry.numbering)) {
+    entry.numbering = validateString(entry.numbering, incrementOptions('numbering', opts));
+  }
+
+  if (defined(entry.id)) {
+    entry.id = validateString(entry.id, incrementOptions('id', opts));
+  }
+
+  if (defined(entry.part)) {
+    entry.part = validateString(entry.part, incrementOptions('part', opts));
+  }
+
+  if (defined(entry.class)) {
+    entry.class = validateString(entry.class, incrementOptions('class', opts));
+  }
+
+  return entry as Entry;
+}
+
+export function validateFileEntry(entry: any, opts: ValidationOptions): FileEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['file'],
+      optional: [...COMMON_ENTRY_KEYS, 'children'],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.file = validateString(outputEntry.file, incrementOptions('file', opts));
+
+  outputEntry = validateCommonEntry(outputEntry, opts);
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  if (defined(entry.children)) {
+    outputEntry.children = validateList(
+      outputEntry.children,
+      incrementOptions('children', opts),
+      (item, ind) => validateEntry(item, incrementOptions(`children.${ind}`, opts)),
+    );
+  }
+
+  return outputEntry as FileEntry;
+}
+
+export function validateURLEntry(entry: any, opts: ValidationOptions): URLEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['url'],
+      optional: [...COMMON_ENTRY_KEYS, 'children'],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.url = validateString(outputEntry.url, incrementOptions('url', opts));
```

Comment:
`validateURLEntry`/`validatePatternEntry`/`validateParentEntry` all follow the same pattern as `validateFileEntry`, so my comments there apply to these as well.